_CS141/part2.md

```diff
@@ -0,0 +1,308 @@
+---
+layout: CS141
+part: true
+title: Lazy Evaluation & Recursion
+---
+
+## Evaluation Strategies
+
+> **Evaluation strategies** determine the order in which we reduce expressions. An expression that can be reduced is called a **redex.**
+
+### Strictness
+
+> A programming language is **strict** if the arguments of a function are evaluated before the function is called.
+
+### Call-by-value
+
+> A **strict evaluation strategy** where all function arguments are reduced to normal forms (values) and then passed as such to the function.
+
+**Example.**
+
+```haskell
+fac' :: Int -> Int -> Int
+fac' 0 m = m
+fac' n m = fac' (n-1) (n*m)
+```
+
+```haskell
+   fac 2
+=> fac' 2 1
+=> fac' (2-1) (2*1)
+=> fac' (1) (2*1)
+=> fac' 1 2
+=> fac' (1-1) (1*2)
+   ...
+```
+
+### Call-by-name
+
+> A **non-strict evaluation strategy** where expressions are given to functions as arguments are not reduced before the function call is made. 
+>
+> Expressions are only reduced when their value is **needed**.
+
+**When is a value needed?**
+
+A case expression is the only thing that enforces the evaluation of a particular expression (we’re only evaluating something if its in a case expression). We **need the value** of an expression when we **cannot proceed** with the case expression **until we reduce** the expression.
+
+```haskell
+fac' :: Int -> Int -> Int
+fac' 0 m = m
+fac' n m = fac' (n-1) (n*m)
+```
+
+```haskell
+   fac 2
+=> fac' 2 1
+=> case 2 of
+     0 -> 1
+     _ -> fac' (2-1) (2*1)
+=> fac' (2-1) (2*1)
+=> case 2-1 of
+     0 -> 2*1
+     _ -> fac' ((2-1)-1) ((2-1)*(2*1))
+=> case 1 of
+     0 -> 2*1
+     _ -> fac' ((2-1)-1) ((2-1)*(2*1))
+```
+
+In **call-by-name**, the 2nd last `=>` to the last `=>` is the only time an expression was evaluated, as the value of the redex `2-1` was needed to continue.
```

Comment:
I don't think this is true: all of steps shown are reductions / there is an expression evaluated at every step.